### ISSUE TITLE
Save model.tar.gz every epoch

### ISSUE THREAD

**Issue by nitishgupta** [allenai/allennlp#1719]
Killing a training process using Ctrl + C, doesn't save the model archive - model.tar.gz even through the code [here](https://github.com/allenai/allennlp/blob/58119c038b07ee5de2c6b0b9f0c6c7d505857ba9/allennlp/commands/train.py#L321) suggests that. 

My code dies with the following messages:
```
pure virtual method called
terminate called without an active exception
Aborted
```
**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.6.6
 - AllenNLP version: v0.6.1
 - PyTorch version: 0.4.1

**Additional context**
As a feature request: can a functionally be added to archive the model whenever the best.th is changed/written instead of only after the training has finished.

**Comment by DeNeutoy** [comment]
I think this may be an issue with your local environment, as you are right that the code you pointed to should create an archive. The error you've reported looks like a compilation time/linking problem in a low level language, perhaps to do with pytorch. Does this only happen with allennlp?

**Comment by nitishgupta** [comment]
Update: This error doesn't always appear. Sometimes is dies with Runtime error on whatever line it was running, but doesn't raise the KeyboardInterrupt exception or archive the model. I tried something similar in a regular python code and it works, so I'm afraid it looks like something to do with allennlp.

Feature request: Can a feature be implemented that the metrics.json and model.tar.gz be written whenever the model is serialized.

**Comment by matt-gardner** [comment]
@nitishgupta, your feature request is already a known issue; see #1645 to follow updates on that.

**Comment by nitishgupta** [comment]
Thanks @matt-gardner

**Comment by matt-gardner** [comment]
That it's not written on interrupt is some kind of environment bug that we probably won't be able to handle.  Writing `model.tar.gz` on every epoch is probably not desirable, as it just adds a bunch of disk space.  It'd be better to have a utility for creating `model.tar.gz` from a serialization directory; though that can be done by hand without too much trouble from an interpreter already.  Either way, it's a separate issue than the one that was reported, and should have its own github issue.  But I wouldn't open that issue ourselves - if a user wants it, they can open the issue.

**Comment by matt-gardner** [comment]
I'm going to close this, as I think the main issue is one we can't fix, and the other issue @nitishgupta brought up is mostly fixed.  If there's more, or something we missed, feel free to re-open this.

**Comment by matt-gardner** [comment]
@nitishgupta, fair enough.  This seems pretty low priority for us at the moment, because the manual work-around exists and works, and this is a pretty infrequent problem.  PRs welcome!

**Comment by nitishgupta** [comment]
Makes sense. I'll try to do this when I get a chance soon. Thanks for considering.

**Comment by bpshaver** [comment]
> Currently. I've written a small bash script to tar manually which works great, but it's still a manual effort.

@nitishgupta Can you share that bash script for those of us in the same boat?

**Comment by nitishgupta** [comment]
```
#!/usr/bin/env

SERIALIZATION_DIR=path_to_weights_directory
WEIGHTS_TH=weights_file  # for eg. best.th

MODEL_ARCHIVE=${SERIALIZATION_DIR}/model.tar.gz

if [ -f "${MODEL_ARCHIVE}" ]; then
    echo "Model archive exists: ${MODEL_ARCHIVE}"
    exit 1
fi

echo "Making model.tar.gz in ${MODEL_ARCHIVE}"

cd ${SERIALIZATION_DIR}

if [ -d "modeltar" ]; then
    echo "modeltar directory exists. Deleting ... "
    rm -r modeltar
fi

mkdir modeltar
cp -r vocabulary modeltar/
cp ${WEIGHTS_TH} modeltar/weights.th
cp config.json modeltar/
cd modeltar
tar -czvf model.tar.gz ./*
cp model.tar.gz ../
cd ../
rm -r modeltar
```

**Comment by bpshaver** [comment]
Thanks @nitishgupta. I actually realized that `allennlp.models.archival.archive_model` does what I need it to do, but you beat me to the punch here.

**Comment by matt-gardner** [comment]
The soon-to-be-merged callback trainer will let you do this pretty easily if you want, so I'm closing this issue.
